tempodb/backend/redis/cache.go

```diff
@@ -0,0 +1,147 @@
+package redis
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk/cache"
+	"github.com/go-kit/kit/log"
+	"github.com/grafana/tempo/tempodb/backend"
+	"github.com/grafana/tempo/tempodb/encoding"
+
+	"github.com/google/uuid"
+)
+
+const (
+	typeBloom = "bloom"
+	typeIndex = "index"
+)
+
+type Config struct {
+	ClientConfig cache.RedisConfig `yaml:",inline"`
+
+	TTL time.Duration `yaml:"ttl"`
+}
+
+type readerWriter struct {
+	nextReader backend.Reader
+	nextWriter backend.Writer
+	client     *cache.RedisCache
+	logger     log.Logger
+}
+
+func New(nextReader backend.Reader, nextWriter backend.Writer, cfg *Config, logger log.Logger) (backend.Reader, backend.Writer, error) {
+	if cfg.ClientConfig.Timeout == 0 {
+		cfg.ClientConfig.Timeout = 100 * time.Millisecond
+	}
+	if cfg.ClientConfig.Expiration == 0 {
+		cfg.ClientConfig.Expiration = cfg.TTL
+	}
+
+	client := cache.NewRedisClient(&cfg.ClientConfig)
+
+	rw := &readerWriter{
+		client:     cache.NewRedisCache("tempo", client, logger),
+		nextReader: nextReader,
+		nextWriter: nextWriter,
+		logger:     logger,
+	}
+
+	return rw, rw, nil
+}
+
+// Reader
+func (r *readerWriter) Tenants(ctx context.Context) ([]string, error) {
+	return r.nextReader.Tenants(ctx)
+}
+
+func (r *readerWriter) Blocks(ctx context.Context, tenantID string) ([]uuid.UUID, error) {
+	return r.nextReader.Blocks(ctx, tenantID)
+}
+
+func (r *readerWriter) BlockMeta(ctx context.Context, blockID uuid.UUID, tenantID string) (*encoding.BlockMeta, error) {
+	return r.nextReader.BlockMeta(ctx, blockID, tenantID)
+}
+
+func (r *readerWriter) Bloom(ctx context.Context, blockID uuid.UUID, tenantID string, shardNum int) ([]byte, error) {
+	key := bloomKey(blockID, tenantID, typeBloom, shardNum)
+	val := r.get(ctx, key)
+	if val != nil {
+		return val, nil
+	}
+
+	val, err := r.nextReader.Bloom(ctx, blockID, tenantID, shardNum)
+	if err == nil {
+		r.set(ctx, key, val)
+	}
+
+	return val, err
+}
+
+func (r *readerWriter) Index(ctx context.Context, blockID uuid.UUID, tenantID string) ([]byte, error) {
+	key := key(blockID, tenantID, typeIndex)
+	val := r.get(ctx, key)
+	if val != nil {
+		return val, nil
+	}
+
+	val, err := r.nextReader.Index(ctx, blockID, tenantID)
+	if err == nil {
+		r.set(ctx, key, val)
+	}
+
+	return val, err
+}
+
+func (r *readerWriter) Object(ctx context.Context, blockID uuid.UUID, tenantID string, start uint64, buffer []byte) error {
+	return r.nextReader.Object(ctx, blockID, tenantID, start, buffer)
+}
+
+func (r *readerWriter) Shutdown() {
+	r.nextReader.Shutdown()
+	r.client.Stop()
+}
+
+// Writer
+func (r *readerWriter) Write(ctx context.Context, meta *encoding.BlockMeta, bBloom [][]byte, bIndex []byte, objectFilePath string) error {
+	for i, b := range bBloom {
+		r.set(ctx, bloomKey(meta.BlockID, meta.TenantID, typeBloom, i), b)
+	}
+	r.set(ctx, key(meta.BlockID, meta.TenantID, typeIndex), bIndex)
+
+	return r.nextWriter.Write(ctx, meta, bBloom, bIndex, objectFilePath)
```

Comment:
Thinking about this now, do we need caching here or is caching at `WriteBlockMeta` below sufficient?

Reply:
I still see caching at both `WriteBlock` and `WriteBlockMeta` but the review is marked `Outdated`. What was changed here?

Reply:
I removed `typeBloom` and `typeIndex`. 

Actually, I thought that you were waiting for Joe. I'm not sure about the pros/cons of only caching at WriteBlockMeta.

Reply:
OK. This PR LGTM except for this question, let me ping @joe-elliott on this.

Reply:
The caching layer should cache bloom and index in the `Write()` method.  It is used by the ingester to flush blocks to the backend.  This way blocks flushed to the backend will be cached as expected for the queriers.

Reply:
Okay spoke to Joe offline, we need this one because there are instances where `WriteBlock` doesn't call `WriteBlockMeta`